crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +286,155 @@ impl Report for PrintJson {
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct PrintSarif {
+    results: Mutex<RefCell<Vec<sarif::Result>>>,
+    error: Mutex<RefCell<Vec<String>>>,
+}
+
+impl Default for PrintSarif {
+    fn default() -> Self {
+        Self {
+            results: Mutex::new(RefCell::new(Vec::new())),
+            error: Mutex::new(RefCell::new(Vec::new())),
+        }
+    }
+}
+
+impl Report for PrintSarif {
+    fn report(&self, msg: Message<'_>) -> Result<(), std::io::Error> {
+        match &msg {
+            Message::Typo(msg) => {
+                let start = String::from_utf8_lossy(&msg.buffer[0..msg.byte_offset]);
+                let column_start =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(start.as_ref(), true).count() + 1;
+                let column_end =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(msg.typo, true).count() + column_start;
+                let line_num = msg.context.as_ref().map(|context| match context {
+                    Context::File(context) => context.line_num,
+                    _ => 1,
+                }).unwrap_or(1);
+                
+                // we have nothing to print here
+                if msg.corrections == typos::Status::Valid {
+                    return Ok(());
+                }
+
+                let message = match &msg.corrections {
+                    typos::Status::Invalid => {
+                        format!("`{}` is disallowed", msg.typo)
+                    }
+                    typos::Status::Corrections(corrections) => {
+                        format!("`{}` should be {}", msg.typo, itertools::join(
+                            corrections.iter().map(|s| format!("`{}`", s)),
+                            ", ",
+                        ))
+                    }
+                    typos::Status::Valid => { unreachable!("has been checked earlier") }
+                };
```

Comment:
Why are we doing two checks on `msg.corrections`?  These should be merged

Reply:
to resolve https://github.com/crate-ci/typos/pull/1047#discussion_r1659498218

Reply:
Thats not what was being asked for there.  That was asking to move the entire `match` up earlier in the function so we would return early on `Valid`, before doing some of the processing that is currently done before `match`.

Reply:
updated

Reply:
ok